### ISSUE TITLE
[FEA] Better handling for nullable columns that do not have any null element

### ISSUE THREAD

**Issue by ttnghia** [rapidsai/cudf#8454]
Currently, many cudf's APIs (such as `copy_bitmask` or `valid_if` and many others) always generate an output bitmask if the input columns are nullable. However, the input columns may not have any null element. Thus, it may be a waste of time and memory to generate such a bitmask containing all valid bit. So, a good practice is to avoid generating a bitmask if the input does not contain any null element. Some APIs have already followed this practice, but not all.

However, if not generating a bitmask for the output, there may be potential issues emerging from the inconsistency between the input and output. Consider an example that an algorithm processing data through several stages. One of the intermediate stages cuts out the bitmask from its output, but all the subsequent stages continue to use the nullable information from the first input. In such cases, the system may crash, or some stages throw an exception, or the final result is undefined. 

So, the question here is, should we keep the consistency between the input and output column? (i.e., if the input column has a bitmask, should we always generate a bitmask for the output column even if it doesn't have any null?)

---------
I have observed a lot of bugs that emerged due to mixing `nullable()` and `null_count()` in the current cudf implementation. Thus, I think a good way to avoid (maybe) all potential issues is to totally remove the `nullable()` API and use `null_count()` all the time. Maybe this is a little bit more expensive for the users to execute the code, but it could definitely eliminate all the headaches from having unexpected bugs and make developers/users much happier.

**Comment by harrism** [comment]
I don't think we should force null counting until we have memoization of the null count.

**Comment by ttnghia** [comment]
Does that mean we already have a plan for implementing null count memoization?

**Comment by harrism** [comment]
Please ignore my comment. Column null_count is already memoized. I was thinking of something else.

**Comment by davidwendt** [comment]
I think the issue is the handling of sliced columns specifically but maybe there are other examples.

If a column does not contain a bitmask, then a sliced column will also not have a bitmask and `copy_bitmask` returns a nullptr bitmask.

If a column does contain a bitmask, then a sliced column will also have a bitmask but the sliced region _may_ technically contain no null entries regardless if the source column contains nulls or not. Here, `copy_bitmask` will create an appropriate sliced copy of the bitmask. Having `copy_bitmask` also check to see if all the copied bits are also all set (non-null) seems unnecessarily expensive.

I think it is reasonable for `copy_bitmask` to return a bitmask if the input `column_view` contains a bitmask without requiring it to also check the bits within the `column_view`'s range.

**Comment by vyasr** [comment]
I interpret `nullable` as "this column may contain nulls" and `has_nulls` (`null_count() > 0`) as "this column currently contains nulls". The former is a semantic distinction about the type of data that is up to the user, whereas the latter is a less fundamental property that only describes the current state of the data. As a consumer of this API I would not expect non-nullable outputs from a `libcudf` function that I called with nullable inputs, except in specific scenarios like reductions where there is some well-defined expected behavior for how nulls are removed. Assuming that you're not proposing to make columns always nullable (which I assume is a non-starter) I wouldn't want nullability to suddenly vanish in the middle of a pipeline.

That said, while I don't think we can optimize away the allocation of the bitmask, parts of libcudf can probably benefit significantly by using `has_nulls` rather than `nullable` to dispatch code paths. By that, I mean that in many cases the only difference between how some function should operate on a non-nullable column and a nullable column with zero nulls is the allocation of the null mask. In cases where that is true, it should be sufficient to simply allocate an all valid null mask and then follow the same code path that would be used for a non-nullable column. I don't know if/how much we leverage that throughout libcudf. Taking advantage of that would not address your concerns about misuse of `has_nulls` vs `nullable` leading to incorrect behavior, but it might help ameliorate performance issues in the case of nullable columns that have no nulls.

**Comment by jrhemstad** [comment]
> These APIs are similar but may yield different/unexpected/wrong results if we don't pay enough attention to their use contexts.

I'm not familiar with a case where the choice of `nullable` vs `has_nulls` will end up in wrong results. I'd think the only difference is in performance. Do you have an example of a bug because of this?

**Comment by jrhemstad** [comment]
> We have that bug not because we implemented the library incorrectly at the beginning. 

My understanding is that that bug was because we weren't passing `column_nullability::FORCE` like was done in https://github.com/rapidsai/cudf/pull/7963
